### PR TITLE
Add BitcoinCash sighash support

### PR DESCRIPTION
**Pull request by junderw**
This should help anyone building apps based off bitcoinjs-lib support Bitcoin Cash.

### CODE REVIEW

**Comment by dcousens** [comment]
Personal NACK,  but will leave open as a reference for others to review.

I don't want this to become a political PR,  the feature is clear,  whether it is useful will depend on events spanning the next few weeks,  and therefore I'd rather defer any action until then.

**Comment by dcousens** [comment]
Needs tests

File: src/transaction.js [hunk 1]
```diff
@@ -251,8 +252,8 @@ Transaction.prototype.clone = function () {
  * hashType, and then hashes the result.
  * This hash can then be used to sign the provided transaction input.
  */
-Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashType) {
-  typeforce(types.tuple(types.UInt32, types.Buffer, /* types.UInt8 */ types.Number), arguments)
+Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashType, inAmount) {
```
Comment (dcousens) [inline src/transaction.js]: Maybe it'd be better to take the approach used for SegWit,  a different function not `hashForSignature`,  avoids positional parameters

**Comment by junderw** [comment]
ToDo:

1. Write tests.
2. Move to separate function
3. Change the fixtures in core/sighash.json to not accidentally trigger 0x40 masking for tests.

I will look into brushing things up once BCC is a clear fixture in the community. For now I'll leave this PR on the back burner.

Thanks for the comments.

File: src/transaction.js (hunk 1 above)
Comment (junderw) [inline src/transaction.js]: I agree.

**Comment by junderw** [comment]
Other problems:

This breaks extractInput for multisigs with partial signatures. As the function for fixing order of sigs requires generating the hash for signing, which means using TransactionBuilder.fromTransaction is not possible.

ie. 3-of-5 multisig rawtx that only contains 2 sigs etc.

File: src/transaction.js [hunk 2]
```diff
@@ -308,10 +309,96 @@ Transaction.prototype.hashForSignature = function (inIndex, prevOutScript, hashT
...
+    var outputsBufOffset = 0
+    outputsBufferArray.forEach(function(outputBuffer) {
+      outputBuffer.copy(outputsBuf, outputsBufOffset)
+      outputsBufOffset += outputBuffer.length
+    })
+
+    // HASH256 to get the hashPrevouts, hashSequence, and hashOutputs
+    var hashPrevouts = bcrypto.hash256(prevOutsBuf)
+    var hashSequence = bcrypto.hash256(inputSequencesBuf)
+    var hashOutputs = bcrypto.hash256(outputsBuf)
```
Comment (afk11) [inline src/transaction.js]: This all comes from BIP143 - could you maybe use the functions from that part instead of duplicating the code again here?

**Comment by jprichardson** [comment]
NACK on my end as well. Couldn't have said it any better than this https://github.com/bitcoinjs/bitcoinjs-lib/pull/826#issuecomment-318262587

**Comment by afk11** [comment]
@junderw @dcousens https://github.com/afk11/bitcoinjs-lib/commit/abd54be36dcf761e0eacd7a132f26cf58f6b34e5

Since the reference implementation performs the call like this (within SignatureHash) it seems appropriate to abstract this logic since bitcoinjs still supports segwit. 

The change as is breaks existing tests (signature hash fixtures came from fuzzing, so are specific to the implementation being exercised (Bitcoin Core without the hard fork). @junderw if you have any test fixtures, I'm sure they'd be appreciated.

Aside: deferring comment on inclusion until we see exactly where this is going.. But I do have a need for this code right now.

**Comment by junderw** [comment]
merged commits from @afk11 via Github UI.

Thanks a bunch.

I agree that this probably shouldn't be merged into master, but rather into another branch maybe... but since most people will come to github and search for "BitcoinCash" under your repo, having this pull request has some meaning imo.

I'll keep my political opinion off here, but I think it helps people of both the anti and pro side to make it easier for BCC users to split off from the community and go do their own thing.

**Comment by afk11** [comment]
@junderw have you got a test case working with this yet? 

trying to sign a transaction (on regtest) and broadcast it but somehow my signature is not correct, added a bunch of debug lines to both bitcoin-abc and bitcoinjs and the signaturehash is the same, just somehow ... the signatures are not!

**Comment by afk11** [comment]
Oops - that patch I introduced to bitcoin-abc was the problem, not my branch xD

https://github.com/afk11/bitcoinjs-lib/tree/hash4cash works for a P2PKH ALL|FORKID. The test case can be reproduced with this:
bitcoin-cli -regtest signrawtransaction 020000000113aaf49280ba92bddfcbdc30d6c7501c2575e4a80f539236df233f9218a2c8400000000000ffffffff0100f2052a010000001976a914cd29cc97826c37281ac61301e4d5ed374770585688ac00000000 '[]' 'null' 'ALL|FORKID'
```
{
  "hex": "020000000113aaf49280ba92bddfcbdc30d6c7501c2575e4a80f539236df233f9218a2c8400000000049483045022100c5874e39da4dd427d35e24792bf31dcd63c25684deec66b426271b4043e21c3002201bfdc0621ad4237e8db05aa6cad69f3d5ab4ae32ebb2048f65b12165da6cc69341ffffffff0100f2052a010000001976a914cd29cc97826c37281ac61301e4d5ed374770585688ac00000000",
  "complete": true
}
```

**Comment by dcousens** [comment]
@junderw close in favour of #831?

**Comment by junderw** [comment]
Yes. Close in favor of #831

**Pull request closed without merging**